### PR TITLE
Compile tee-worker against rust-sgx-sdk v2.0

### PR DESCRIPTION
**Pull request by Kailai-Wang**
### Context

After a few weeks' intermittent work, finally, we may rejoice.

> [!IMPORTANT]
> This PR is ready to be reviewed and merged, but I suggest that we wait until state provisioning is tested on some host. 
> 
> In theory, the change in this PR doesn't affect the client, but I label it as `C1` to mark as a hint to any unexpected error (e.g. SEGV) that might arise.

### Why do we need a new SDK

The main reasons are:
- **blocking**: newer polkadot/substrate dependencies need a newer rustc version, which isn't supported by the current sdk (v1.1.6). It blocks our further version update.
- **non-blocking but nice-to-have**: some crates (e.g. `serde_json`) have now `no_std` support, however, they need a newer rustc too. It's preferred to use no-std crate over ported sgx version of crates.
- sdk-v2.0 provides other benefits, please check [their README](https://github.com/apache/incubator-teaclave-sgx-sdk/blob/v2.0.0-preview/README.md)

### SDK

I'm using [v2.0.0-preview](https://github.com/apache/incubator-teaclave-sgx-sdk/tree/v2.0.0-preview) branch of rust sgx sdk, this branch is not "officially" released but has been there for a while. It has been sporadically worked on, the unittests and samplecode are provided too - I tested most of them - looking good.

I do need to make some small adjustments in [my own fork](https://github.com/Kailai-Wang/incubator-teaclave-sgx-sdk/tree/v2.0.0-preview), please check TOML files for overriding:
- `.cpp` file compiles under ubuntu20, but not on ubutu22 due to a different (newer) gcc verison
- [pointer misalignment fix](https://github.com/Kailai-Wang/incubator-teaclave-sgx-sdk/pull/1)

I'm considering submitting them as upstream PRs.
The associated rust-toolchain version is `nightly-2023-11-17`, but I believe it's possible to further upgrade it.

### Other SDK-related files

I basically follow the example code in the sdk-v2.0 for the following files:
- Makefile (except I don't touch anything regarding `BUILD_STD`)
- Enclave configuration
- *.edl files and inc for enclave

### Other SGX ported crates

I need to manually fork a bunch of sgx-ported crates to pair with the sdk-v2.0 (mainly `sgx_tstd` dep): https://github.com/Kailai-Wang?tab=repositories&q=sgx

They are under my personal workspace for now but can be migrated to `litentry/` org after we obversing that they work well.

Among those crates, `once_cell` is the (only?) one that has `no_std` support but I still use the ported version, because it demands some `critical-section` implementation, which is unclear under sgx - I haven't found any either.

> [!NOTE]
> The following are the problems / issues that I've met during the re-anchoring, some are resolved, some are work-arounded, some still need further investigation.

### Intel SGX-SDK version

I have to use Intel sgx-sdk version 2.17, otherwise you'll get unmatched error when signing the enclave, see https://github.com/apache/incubator-teaclave-sgx-sdk/issues/387

But in general, it should be possible to adjust it for newer sdk - we can have follow-up issues. **We also need to test HW mode on a real host (preferrably ubuntu22.04) to see how it goes**. Both CI and my local tests are conducted in SW mode.

> [!NOTE]
> it's now compiled against Intel SDK 2.21: https://github.com/litentry/litentry-parachain/pull/2813

### De-/serialization of `Rsa3072KeyPair`
In sdk-v2.0, it implements `sgx_serialize::{Deserialize, Serialize}`. I was thinking about replacing it with no_std `serde` but eventually ditched it, mainly because:
- it seems unfavorable to depend on third party crates within the sdk
- `sgx_serialize` is widely used in sdk for other structs too, so I didn't mess around with it (in case others depend on the de-/serialization result of it)

As a result, I use `sgx_serialize::json::{encode, decode}` to deal with Rsa3072KeyPair de/ser. It should produce the same result as before and our ts-test should prove that.

### SgxFile could be unstable
SgxFile is moved to `sgx_tprotected_fs::SgxFile`, there's a new `HostFile` struct. These seem to incur some problems with SgxFile creation/open/read:
- pointer missalignment in `u_sgxfs_open_ocall` - see fix above
- Previously generated sealed files can't be read directly by the new enclave - even for the same MRSIGNER and on the same host. I get 
```
[ERROR enclave_runtime::error] Returning error Crypto(IO(Os { code: 2, kind: NotFound, message: "No such file or directory" })) as sgx unexpected.
```
even though the file is there. This must have something to do with the new `HostFile` impl. We can further look into this, but we should always be able to provision the state via RPC.

### Use of `USE_OPT_LIBS ?= 0`
I have to use `USE_OPT_LIBS ?= 0`, otherwise `sgx_crypto_sys` and `sgx_tlibc_sys` complain prebuilt libs can't be found.
There's `optimized_libs/external/libirc/efi2/libirc.a` but not `libm*` => maybe we can further improve this.

### Use of `SGX_DEBUG=1`
As you can see in Makefile, I use `SGX_DEBUG=1` for the enclave compilation, and strip the symbols in the end to reduce the .so size. If compiled with `SGX_DEBUG=0` - which basically implies `release` profile for `cargo` - the enclave will emit a SEGV during initialization, I've narrowed it down to [this line](https://github.com/apache/incubator-teaclave-sgx-sdk/blob/2d1e0a43e996dff95d0b33010496536704d58e85/sgx_protected_fs/tfs/src/sys/keys.rs#L137): 
```
report: Box::new(*Report::get_self()),
```
I didn't look into it as it could lie in somewhere deep (like compiler or fundamental crates/impl). Maybe it will be self-resolved as we upgrade them, but I can ask in upstream repo again

### Use of `cargo update`
Most crates can be upgraded with a few exceptions:
- `rustls-pki-types` => has to be pinned at `0.2.1`
- `substrate-bip39` => has to be pinned at `0.4.4`

They could be self-resolved when we upgrade the crate/tc version.

> [!TIP]
> So given that 371 files are changed, how to review this PR?

Mainly these files:
- `Makefile`
- `TOML` files, especially `tee-worker/Cargo.toml` and `tee-worker/enclave-runtime/Cargo.toml`
- `.rs` files with non-trivial changes, by "trivial" I mean renaming like `sgx_status_t` => `SgxStatus`.

### CODE REVIEW

**Comment by Kailai-Wang** [comment]
It's not guaranteed that I wrote out all noticeable things - if I have something later I'll add it.

File: tee-worker/core/rpc-client/src/direct_client.rs [hunk 1]
```diff
@@ -174,7 +172,8 @@ impl DirectApi for DirectClient {
 		let response_str = self.get(&jsonrpc_call)?;
 
 		let shielding_pubkey_string = decode_from_rpc_response::<String>(&response_str)?;
-		let shielding_pubkey: Rsa3072PubKey = serde_json::from_str(&shielding_pubkey_string)?;
+		let shielding_pubkey: Rsa3072PublicKey =
+			sgx_serialize::json::decode(&shielding_pubkey_string).unwrap();
```
Comment (kziemianek) [inline tee-worker/core/rpc-client/src/direct_client.rs]: can we avoid unwrap here ?

File: tee-worker/enclave-runtime/Enclave.config.production.xml [hunk 2]
```diff
@@ -2,11 +2,18 @@
...
+  <ReservedMemMinSize>0x20000</ReservedMemMinSize>
+  <ReservedMemInitSize>0x40000</ReservedMemInitSize>
+  <ReservedMemMaxSize>0x400000</ReservedMemMaxSize>
+  <TCSMaxNum>1024</TCSMaxNum>
   <TCSNum>1024</TCSNum>
-  <TCSPolicy>0</TCSPolicy> <!-- 0 = Thread Control Structure (TCS) is bound to the untrusted thread -->
-  <DisableDebug>1</DisableDebug>
+  <TCSMinPool>64</TCSMinPool>
+  <TCSPolicy>0</TCSPolicy>
+  <DisableDebug>0</DisableDebug>
```
Comment (kziemianek) [inline tee-worker/enclave-runtime/Enclave.config.production.xml]: Shouldn't we keep debug disabled for prod ?

**Review by kziemianek — APPROVED** [review]
Rust code looks fine, nice job.

File: tee-worker/enclave-runtime/Enclave.config.production.xml (hunk 2 above)
Comment (Kailai-Wang) [reply tee-worker/enclave-runtime/Enclave.config.production.xml]: Good catch, thanks!

File: tee-worker/core/rpc-client/src/direct_client.rs (hunk 1 above)
Comment (Kailai-Wang) [reply tee-worker/core/rpc-client/src/direct_client.rs]: Fixed

**Comment by Kailai-Wang** [comment]
~~2.21 SDK update is blocked by https://github.com/apache/incubator-teaclave-sgx-sdk/issues/456 (mainly CLI tools)~~

**Comment by silva-fj** [comment]
I noticed that on this branch rust-analyzer crashes with the following error:

```bash
2024-06-24T10:11:23.776955Z ERROR rust_analyzer::main_loop: FetchWorkspaceError:
rust-analyzer failed to load workspace: Failed to load the project at /home/franjs/projects/litentry-parachain/tee-worker/Cargo.toml: Failed to read Cargo metadata from Cargo.toml file /home/franjs/projects/litentry-parachain/tee-worker/Cargo.toml, Some(Version { major: 1, minor: 76, patch: 0, pre: Prerelease("nightly") }): Failed to run `cd "/home/franjs/projects/litentry-parachain/tee-worker" && RUSTUP_TOOLCHAIN="/home/franjs/.rustup/toolchains/nightly-2023-11-17-x86_64-unknown-linux-gnu" SKIP_WASM_BUILD="1" "/home/franjs/.cargo/bin/cargo" "metadata" "--format-version" "1" "--all-features" "--manifest-path" "/home/franjs/projects/litentry-parachain/tee-worker/Cargo.toml" "--filter-platform" "x86_64-unknown-linux-gnu"`: `cargo metadata` exited with an error: warning: skipping duplicate package `server` found at `/home/franjs/.cargo/git/checkouts/incubator-teaclave-sgx-sdk-e1a16a9ae7d3e68c/905cfba/samplecode/rpc/server/app`
warning: skipping duplicate package `client` found at `/home/franjs/.cargo/git/checkouts/incubator-teaclave-sgx-sdk-e1a16a9ae7d3e68c/905cfba/samplecode/rpc/client/app`
warning: skipping duplicate package `app` found at `/home/franjs/.cargo/git/checkouts/incubator-teaclave-sgx-sdk-e1a16a9ae7d3e68c/905cfba/samplecode/hyper-rustls-https-server/app`
warning: skipping duplicate package `app` found at `/home/franjs/.cargo/git/checkouts/incubator-teaclave-sgx-sdk-e1a16a9ae7d3e68c/905cfba/samplecode/cov/app`
warning: skipping duplicate package `app` found at `/home/franjs/.cargo/git/checkouts/incubator-teaclave-sgx-sdk-e1a16a9ae7d3e68c/905cfba/samplecode/regex/app`
warning: skipping duplicate package `app` found at `/home/franjs/.cargo/git/checkouts/incubator-teaclave-sgx-sdk-e1a16a9ae7d3e68c/905cfba/samplecode/zlib-lazy-static-sample/app`
warning: skipping duplicate package `app` found at `/home/franjs/.cargo/git/checkouts/incubator-teaclave-sgx-sdk-e1a16a9ae7d3e68c/905cfba/samplecode/template/app`
warning: skipping duplicate package `app` found at `/home/franjs/.cargo/git/checkouts/incubator-teaclave-sgx-sdk-e1a16a9ae7d3e68c/905cfba/samplecode/crypto/app`
warning: skipping duplicate package `app` found at `/home/franjs/.cargo/git/checkouts/incubator-teaclave-sgx-sdk-e1a16a9ae7d3e68c/905cfba/samplecode/seal/app`
warning: skipping duplicate package `app` found at `/home/franjs/.cargo/git/checkouts/incubator-teaclave-sgx-sdk-e1a16a9ae7d3e68c/905cfba/samplecode/helloworld/app`
warning: skipping duplicate package `app` found at `/home/franjs/.cargo/git/checkouts/incubator-teaclave-sgx-sdk-e1a16a9ae7d3e68c/905cfba/samplecode/httpreq/app`
warning: skipping duplicate package `app` found at `/home/franjs/.cargo/git/checkouts/incubator-teaclave-sgx-sdk-e1a16a9ae7d3e68c/905cfba/samplecode/mutual-ra/app`
warning: skipping duplicate package `app` found at `/home/franjs/.cargo/git/checkouts/incubator-teaclave-sgx-sdk-e1a16a9ae7d3e68c/905cfba/samplecode/switchless/app`
warning: skipping duplicate package `app` found at `/home/franjs/.cargo/git/checkouts/incubator-teaclave-sgx-sdk-e1a16a9ae7d3e68c/905cfba/samplecode/logger/app`
warning: skipping duplicate package `Helloworldsampleenclave` found at `/home/franjs/.cargo/git/checkouts/log-sgx-5c3c9a7506bc589e/64e1bc5/sgx/log-sgx-test/enclave`
warning: skipping duplicate package `app` found at `/home/franjs/.cargo/git/checkouts/log-sgx-5c3c9a7506bc589e/64e1bc5/sgx/log-sgx-test/app`
warning: skipping duplicate package `Helloworldsampleenclave` found at `/home/franjs/.cargo/git/checkouts/rust-memchr-sgx-36763dcc50cf9a1d/fb51ee3/sgx/memchr-sgx-test-libc/enclave`
warning: skipping duplicate package `app` found at `/home/franjs/.cargo/git/checkouts/rust-memchr-sgx-36763dcc50cf9a1d/fb51ee3/sgx/memchr-sgx-test-libc/app`
warning: skipping duplicate package `Helloworldsampleenclave` found at `/home/franjs/.cargo/git/checkouts/log-sgx-b1b3718c01599529/2ca9039/sgx/log-sgx-test/enclave`
warning: skipping duplicate package `app` found at `/home/franjs/.cargo/git/checkouts/log-sgx-b1b3718c01599529/2ca9039/sgx/log-sgx-test/app`
error: the crate `sgx_crypto v2.0.0 (https://github.com/Kailai-Wang/incubator-teaclave-sgx-sdk?branch=v2.0.0-sdk-2.21#905cfba0)` depends on crate `sgx_serialize v2.0.0 (https://github.com/Kailai-Wang/incubator-teaclave-sgx-sdk?branch=v2.0.0-sdk-2.21#905cfba0)` multiple times with different names

```

**Comment by Kailai-Wang** [comment]
We can close this PR for now as it's not urgently needed after we were able to separate substrate deps between parachain and worker.

But let's not delete the branch - we might want to revisit it (if we go for 2.0)

**Pull request closed without merging**